nixos/systemd: add an option for ExecCondition

danth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

The other `Exec*` options in SystemD already have NixOS equivalents, so it makes sense to add this one for consistency.

A description of what `ExecCondition` does can be found [here](https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecCondition=).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/633

SuperSandro2000 commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -294,6 +294,10 @@ let
               makeJobScript "${name}-post-stop" config.postStop;
           })
+        (mkIf (config.condition != "")
+          { serviceConfig.ExecCondition =
```
Please sort this alphabetical.

SuperSandro2000 commented on nixos/modules/system/boot/systemd-unit-options.nix:
```diff
@@ -321,6 +321,15 @@ in rec {
+      type = types.lines;
+      default = "";
+      description = ''
+        Shell commands executed to determine whether the service
```
This can be written without the line break.

Ma27 commented:
Does this have any benefits over `serviceConfig.ExecCondition`?

The thing is, `script` is e.g. only a historical artifact from a past where NixOS used upstart. Also, it's IIMHO not a nice thing to use these options since most things are one-liners (that can be split over multiple lines in `.service`-files) that are now written to its own store-path which introduces another level of indirection. This means that when I e.g. run `systemctl cat foo.service`, I only see a store-path for `Exec*`-settings rather than what's actually happening.

In all other cases I don't see what's wrong with `${pkgs.writeShellScript "foo" "bar"}`.

SuperSandro2000 commented:
> The thing is, script is e.g. only a historical artifact from a past where NixOS used upstart.

Off topic question but should we remove the other execs and redirect them to serviceConfig?

danth commented:
> Does this have any benefits over `serviceConfig.ExecCondition`?

This and the other script options are passed through `makeJobScript`:

https://github.com/NixOS/nixpkgs/blob/a97f20e5d822a373153dca35058462dc00a0cf5b/nixos/modules/system/boot/systemd.nix#L209-L226

Scripts are written into a sub-directory to avoid cluttering the log with lengthy derivation names.

To replicate this behaviour using only `serviceConfig.ExecCondition`, you would need to write a complex expression like `${"${pkgs.writeShellScriptBin "foo" "bar"}/bin/foo"}`. This is especially difficult to read when bar is a multi-line string.

> Also, it's IIMHO not a nice thing to use these options since most things are one-liners

Would it be possible for `makeJobScript` to return one-liners verbatim while still writing a file for longer scripts?

danth committed:
nixos/systemd: add an option for ExecCondition

The other Exec* options in SystemD already have NixOS equivalents, so it
makes sense to add this one for consistency.
ExecCondition is like an ExecStartPre and condition check hybrid.
